src/protocols/secure_channel/CASESession.cpp

```diff
@@ -226,6 +240,12 @@ class CASESession::WorkHelper
         auto status = DeviceLayer::PlatformMgr().ScheduleWork(AfterWorkHandler, reinterpret_cast<intptr_t>(helper));
         if (status != CHIP_NO_ERROR)
         {
+            // We failed to schedule after work callback, so setting mScheduleAfterWorkFailed flag to true
+            // This can be checked from foreground thread and after work callback can be retried
+            ChipLogError(SecureChannel, "Failed to Schedule the AfterWorkCallback on foreground thread");
+            helper->mScheduleAfterWorkFailed.store(true);
+            helper->mStatus = status;
+
             // Release strong ptr since scheduling failed
             helper->mStrongPtr.reset();
```

Comment:
Please double-check me carefully on this @shubhamdp and @mlepage-google: now that we have this watchdog mechanism, we should NOT reset helper->mStrongPtr here.

Here's why: let's say we reset it (and to be safe we reset it before our mScheduleAfterWorkFailed synchronization point).  At that point the only reference to us is the one the session is holding.  When we now call the after-work handler it will call into the session, and the session has to be careful to not drop its ref before it's done using the data (which is owned by this object).  Right now that happens to be true, but it's very fragile.

The reason we had this reset bit was to avoid leaking due to this ref that would never go away, but now we _do_ have a mechanism for it to go away in the form of the watchdog, right?

Reply:
I think it's fine to reset the helper's strong ptr to itself here. From its perspective, it's done. The session does still hold a strong ptr to the helper, and yes it has to properly manage that as it does anything with the helper.

If the CASESession were destroyed (object deleted), the helper would still need to be deleted.

At the same time, we can't put ownership solely in the CASESession, because it can't clean up if a scheduled function is waiting to run.

Reply:
> and yes it has to properly manage that as it does anything with the helper

Well, SendSigma3c does not manage it properly in that case....  we should just do a followup to:
1. Document the expectations on CASESession clearly for these ownership bits.
2. Ensure that those expectations are followed.

Reply:
I have filed https://github.com/project-chip/connectedhomeip/issues/27541 for tracking this.

Reply:
Can we resolve the conversation as the follow-up issue is filed? @bzbarsky-apple @mlepage-google

Reply:
Resolving this thread!